### PR TITLE
Development

### PR DESCRIPTION
**Pull request by looselycoupled**
For internal code review.  Functionality is complete per roadmap for this version however latest version is still missing some comments and tests.  This will be updated in a future minor release.

### CODE REVIEW

File: btrdb/conn.py [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+class BTrDB(object):
+    """
+
+    """
+
+    def __init__(self, endpoint):
+        self.ep = endpoint
+
+
+    def streams(self, *identifiers, versions=[]):
```
Comment (looselycoupled) [inline btrdb/conn.py]: remove literal and use None instead

File: btrdb/conn.py [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+
+        versions: list[int]
+            a single or iterable of version numbers to match the identifiers
+
+        """
+        if versions and len(versions) != len(identifiers):
+            raise ValueError("Number of versions does not match identifiers")
+
+        if not versions:
+            versions = [0] * len(identifiers)
```
Comment (looselycoupled) [inline btrdb/conn.py]: what happens to versions?

File: btrdb/stream.py [hunk 3]
```diff
@@ -0,0 +1,893 @@
...
+##########################################################################
+## Utility Classes
+##########################################################################
+
+class StreamFilter(object):
+    """
+    Object for storing requested filtering options
+    """
+    def __init__(self, start=None, end=None):
+        self.start = int(start) if start else None
```
Comment (looselycoupled) [inline btrdb/stream.py]: make a util function for `to_ns` which accepts int, float, datetime, numpy.datetime64, string (parseable date)

File: btrdb/stream.py [hunk 4]
```diff
@@ -0,0 +1,893 @@
...
+        if self._known_to_exist:
+            return True
+
+        try:
+            self.refresh_metadata()
+            return True
+        except BTrDBError as bte:
+            if bte.code == 404:
+                return False
+            raise
```
[the remark this answers is not in the document]
Comment (looselycoupled) [reply btrdb/stream.py]: It would be great to enumerate all of the possible codes (add to Jira)

File: btrdb/stream.py [hunk 5]
```diff
@@ -0,0 +1,893 @@
...
+        version = 0
+        while i < len(data):
+            thisBatch = data[i:i + INSERT_BATCH_SIZE]
+            version = self._btrdb.ep.insert(self._uuid, thisBatch)
+            i += INSERT_BATCH_SIZE
+        return version
+
+    def _update_tags_collection(self, tags, collection):
+        tags = self.tags() if tags is None else tags
+        collection = self.collection() if collection is None else collection
```
Comment (looselycoupled) [inline btrdb/stream.py]: potential bug if collection was never cached.  could alleviate by raising error if collection is None

File: btrdb/stream.py [hunk 6]
```diff
@@ -0,0 +1,893 @@
...
+        Raises
+        ------
+        BTrDBError [401] no such point
+            No point satisfies the query in the direction specified
+
+        """
+
+        rp, version = self._btrdb.ep.nearest(self._uuid, time, version, backward)
+        return RawPoint.from_proto(rp), version
+
```
Comment (looselycoupled) [inline btrdb/stream.py]: can we add `earliest` and `latest` for single streams?

File: btrdb/stream.py [hunk 7]
```diff
@@ -0,0 +1,893 @@
...
+        windows, this accuracy may not be required. For example for a window of
+        a day, +- one second may be appropriate, so a depth of 30 can be
+        specified. This is much faster to execute on the database side.
+
+        """
+        if not self.allow_window:
+            raise InvalidOperation("A window operation is already requested")
+
+        # TODO: refactor keeping in mind how exception is raised
+        self.width = int(width)
```
Comment (looselycoupled) [inline btrdb/stream.py]: Create github issues for better casting/conversion to `int`

File: btrdb/stream.py [hunk 8]
```diff
@@ -0,0 +1,893 @@
...
+        return params
+
+    @property
+    def values_iter(self):
+        """
+        Must return context object which would then close server cursor on __exit__
+        """
+        raise NotImplementedError()
+
+    @property
```
Comment (looselycoupled) [inline btrdb/stream.py]: Requested to take out property here since this is a roundtrip to a server

File: btrdb/stream.py [hunk 9]
```diff
@@ -0,0 +1,893 @@
...
+    def values(self):
+        """
+        Returns a fully materialized list of lists for the stream values/points
+        """
+        result = []
+        params = self._params_from_filters()
+        stream_output_iterables = [s.values(**params) for s in self._streams]
+
+        for stream_output in stream_output_iterables:
```
Comment (looselycoupled) [inline btrdb/stream.py]: dont create all the generators at once as each one is a server side cursor

File: btrdb/conn.py (hunk 1 above)
Comment (looselycoupled) [reply btrdb/conn.py]: done

**Comment by looselycoupled** [comment]
import `uuid` as `uuidlib` in order to change `uu` argument/variables to `uuid` to be more descriptive.

**Comment by immesys** [comment]
Can I close this now that you have #16?

**Comment by looselycoupled** [comment]
This PR was used purely for code review and has been superseded by PR #16

**Pull request closed without merging**